language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.javaapi.data.codegen;
+
+import com.daml.ledger.javaapi.data.CreatedEvent;
+import com.daml.ledger.javaapi.data.DamlRecord;
+import com.daml.ledger.javaapi.data.Identifier;
+import com.daml.ledger.javaapi.data.Value;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+public abstract class ContractCompanion<Ct, Id, Data> {
+  public final Identifier TEMPLATE_ID;
+  final String templateClassName; // not something we want outside this package
+
+  protected final Function<String, Id> newContractId;
+  protected final Function<DamlRecord, Data> fromValue;
+
+  @Deprecated
+  public abstract Ct fromIdAndRecord(String contractId, DamlRecord record$);
```

Comment:
Would it make sense not to add an immediately deprecated method?

Reply:
We want the new concrete methods to be deprecated, as they are only there to support existing deprecated methods that are now static forwarders (removing all of these can be a separate issue), but we can

* [ ] remove the deprecated abstract parent

Reply:
0de313ed55743a0e0c9b980b627450d314aaa5a2